File: code/algo/getoffer/go/2/fullbst_insert.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type CBTInserter struct {
	root      *TreeNode
	candidate []*TreeNode
}

func NewCBT(root *TreeNode) CBTInserter {
	q := []*TreeNode{root}
	candidate := []*TreeNode{}
	for len(q) > 0 {
		node := q[0]
		q = q[1:]
		if node.Left != nil {
			q = append(q, node.Left)
		}
		if node.Right != nil {
			q = append(q, node.Right)
		}
		if node.Left == nil || node.Right == nil {
			candidate = append(candidate, node)
		}
	}
	return CBTInserter{root, candidate}
}

func (c *CBTInserter) Insert(val int) int {
	child := &TreeNode{Val: val}
	node := c.candidate[0]
	if node.Left == nil {
		node.Left = child
	} else {
		node.Right = child
		c.candidate = c.candidate[1:]
	}
	c.candidate = append(c.candidate, child)
	return node.Val
}

func (c *CBTInserter) Get_root() *TreeNode {
	return c.root
}

File: code/algo/getoffer/c++/2/fullbst_insert.cpp
#include <queue>
#include <vector>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class CBTInserter {
public:
  CBTInserter(TreeNode *root) : root(root) {
    std::queue<TreeNode *> q;
    q.push(root);
    while (!q.empty()) {
      TreeNode *node = q.front();
      q.pop();
      if (node->left) {
        q.push(node->left);
      }
      if (node->right) {
        q.push(node->right);
      }
      if (!node->left || !node->right) {
        candidate.push_back(node);
      }
    }
  }

  int insert(int val) {
    TreeNode *child = new TreeNode(val);
    TreeNode *node = candidate.front();
    if (!node->left) {
      node->left = child;
    } else {
      node->right = child;
      candidate.erase(candidate.begin());
    }
    candidate.push_back(child);
    return node->val;
  }

  TreeNode *get_root() { return root; }

private:
  TreeNode *root;
  std::vector<TreeNode *> candidate;
};
